src/shared/dissect-image.c

```diff
@@ -1270,27 +1321,60 @@ static int verity_partition(
+                        return r;
+                /* libdevmapper can return EINVAL when the device is already in the activation stage.
+                 * There's no way to distinguish this situation from a genuine error due to invalid
+                 * parameters, so immediately fallback to activating the device with a unique name.
+                 * Improvements in libcrypsetup can ensure this never happens: https://gitlab.com/cryptsetup/cryptsetup/-/merge_requests/96 */
+                if (r == -EINVAL && FLAGS_SET(flags, DISSECT_IMAGE_VERITY_SHARE))
+                        return verity_partition(m, v, root_hash, root_hash_size, verity_data, NULL, root_hash_sig ?: hash_sig_from_file, root_hash_sig_size, flags & ~DISSECT_IMAGE_VERITY_SHARE, d);
+                if (r == -EEXIST) {
+                        if (!restore_deferred_remove){
+                                /* To avoid races, disable automatic removal on umount while setting up the new device. Restore it on failure. */
+                                r = dm_deferred_remove_cancel(name);
+                                if (r < 0)
+                                        return log_error_errno(r, "Disabling automated deferred removal for verity device %s failed: %m", node);
+                                restore_deferred_remove = strdup(name);
+                                if (!restore_deferred_remove)
+                                        return log_oom();
+                        }
+
+                        r = verity_can_reuse(root_hash, root_hash_size, !!root_hash_sig || !!hash_sig_from_file, name, &cd);
+                        if (!IN_SET(r, 0, -ENODEV, -ENOENT)) {
+                                return log_error_errno(r, "Checking whether existing verity device %s can be reused failed: %m", node);
+                        }
+                }
+                if (r == 0) {
+                        /* Sanity check: libdevmapper is known to report that the device already exists and is active,
+                         * but it's actually not there, so the later filesystem probe or mount would fail. */
+                        r = access(node, F_OK);
+                        break;
+                }
+        }
+
+        /* An existing verity device was reported by libcryptsetup/libdevmapper, but we can't use it at this time.
+         * Fall back to activating it with a unique device name. */
+        if (r != 0 && FLAGS_SET(flags, DISSECT_IMAGE_VERITY_SHARE))
+                return verity_partition(m, v, root_hash, root_hash_size, verity_data, NULL, root_hash_sig ?: hash_sig_from_file, root_hash_sig_size, flags & ~DISSECT_IMAGE_VERITY_SHARE, d);
+
+        /* Everything looks good and we'll be able to mount the device, so deferred remove will be re-enabled at that point. */
+        restore_deferred_remove = mfree(restore_deferred_remove);
+
+        if (!GREEDY_REALLOC0(d->decrypted, d->n_allocated, d->n_decrypted + 1))
+                return -ENOMEM;
```

Comment:
i'd probably do this early on (before the loop), since if this succeeds but the other stuff then does not it has no bad effects, but in this current order that's not really the case...

Reply:
I'll move it back. The thinking was, in theory if cryptsetup operations fail, it should have no visible side effects, but this would waste memory. On the other hand there could be strange failure modes, and this realloc'ed memory would just be reused next time around, so it does sound safer overall.